src/object.d

```diff
@@ -2309,6 +2310,85 @@ inout(V) get(K, V)(inout(V[K])* aa, K key, lazy inout(V) defaultValue)
     }
 }
 
+unittest
+{
+    static class T
+    {
+        static size_t count;
+        this() { ++count; }
+    }
+
+    T[string] aa;
+
+    auto a = new T;
+    aa["foo"] = a;
+    assert(T.count == 1);
+    auto b = aa.get("foo", new T);
+    assert(T.count == 1);
+    assert(b is a);
+    auto c = aa.get("bar", new T);
+    assert(T.count == 2);
+    assert(c !is a);
+
+    //Obviously get doesn't add.
+    assert("bar" !in aa);
+}
+
+auto getOrAdd(K, V)(ref V[K] aa, K key, lazy V value)
```

Comment:
Return by `ref`, please. Could also use `V.init` as a default for `value`. Then we can write code like this:
```d
struct S { int value; }
S[string] aa;
aa.getOrAdd("foo").value = 42;
assert(aa == ["foo": S(42)]);
```

Reply:
I wondered whether it should be `ref` (which is why I added the basic value unit test), I hadn't considered this scenario, I will add a test for it and update.

Reply:
The default value is also a nice idea, so I will add that too (I didn't know lazy args could have a default value)

Reply:
Hm... I just suggested a name of `getPtr` on the newsgroup, as I assumed that's what you were doing, but it seems you actually are returning by value. I think returning by ref is preferable to either (maybe `getRef`?). Note that if you wanted to create a pointer to the value, you have to go through a *triple* lookup today:

```D
Value *p = void;
if(!(p = key in aa)) // one
{
    aa[key] = initializeValue(); // two
    p = key in aa;  // three
}
```
Having this function return by ref makes this very easy to do in one lookup.

Note that a class value type is going to be confusing here unless you do a default value of `new Value`.

Reply:
@schveiguy  `getPtr`and `getRef` are certainly both terse which I like. My only criticism is that they don't as clearly communicate that the value will be added/inserted into the associative array.

I am a little bit confused by the triple lookup example. Can I not just do `p = (aa[key] = initializeValue())`

That said this PR will make this much more simple: `Value* p = aa.getOrAdd(key, initializeValue());` Is there a unittest you can suggest to ensure this works as you expected?

Reply:
> Can I not just do `p = (aa[key] = initializeValue())`

You are still thinking in terms of classes. You may want a pointer to a value type (think `int[int]`), where you need to use `key in aa` to avoid further lookups.

Reply:
>In all honesty, before looking closer, I thought the get function already did this.
Ha, yeah I made the same mistake...that lead to a few hours of head scratching ;)

Reply:
@schveiguy I have added a unittest which I think covers your example. @aG0aep6G I've made the updates as you've suggested too.